feature: add FFI interface to verify SSL client certificate

ArchangelSDY opened the pull request:
This PR adds a FFI interface that allows users to configure SSL client certificate verification dynamically. For example, Nginx can now read SNI first and then determine whether to request a client certificate during handshake. Optionally, caller can pass in a CA list used for verification.

I hereby granted the copyright of the changes in this pull request
to the authors of this lua-nginx-module project.

doujiang24 commented:
@ArchangelSDY will you create another PR in `lua-resty-core` to expose Lua API?

ArchangelSDY commented:
@doujiang24  Sure if you are fine with the API change in this PR. Here they are:

https://github.com/openresty/lua-resty-core/pull/289
https://github.com/openresty/meta-lua-nginx-module/pull/75
https://github.com/openresty/stream-lua-nginx-module/pull/190

doujiang24 commented on src/ngx_http_lua_ssl_certby.c:
```diff
@@ -1294,4 +1294,104 @@ ngx_http_lua_ffi_set_priv_key(ngx_http_request_t *r,
+
+            /* add subject to name chain, which will be sent to client */
+            if ((subject = X509_NAME_dup(
+                    X509_get_subject_name(x509))) == NULL) {
```
style: we usually use two lines here.
```suggestion
    subject = X509_NAME_dup(X509_get_subject_name(x509))
    if (subject == NULL) {
```

doujiang24 commented on t/140-ssl-c-api.t:
```diff
@@ -12,7 +12,7 @@ if ($openssl_version =~ m/built with OpenSSL (0|1\.0\.(?:0|1[^\d]|2[a-d]).*)/) {
 
 } else {
-    plan tests => repeat_each() * (blocks() * 5 + 1);
+    plan tests => repeat_each() * 38;
```
style: we'd better keep this style: `block() * 5 + N`.

ArchangelSDY replied:
Sure. Reverted back. It took me a while to figure out what does `5` and `N` mean :)

doujiang24 commented:
@ArchangelSDY we'd better use new commit for the new changes, force push is not friendly for the reviewer to check the new changes only, thanks :)

ArchangelSDY commented:
I see. Will follow that for new changes.

doujiang24 commented on src/ngx_http_lua_ssl_certby.c:
```diff
@@ -1294,4 +1294,104 @@ ngx_http_lua_ffi_set_priv_key(ngx_http_request_t *r,
+
+int
+ngx_http_lua_ffi_ssl_verify_client(ngx_http_request_t *r,
+    int depth,
```
style: this line can move to the suffix of the previous line.

doujiang24 commented on src/ngx_http_lua_ssl_certby.c:
```diff
@@ -1294,4 +1294,104 @@ ngx_http_lua_ffi_set_priv_key(ngx_http_request_t *r,
+int
+ngx_http_lua_ffi_ssl_verify_client(ngx_http_request_t *r,
+    int depth,
+    void *cdata, char **err)
```
`cdata` may not be a good name here, better use `ca_chain` instead?

doujiang24 commented on src/ngx_http_lua_ssl_certby.c:
```diff
@@ -1294,4 +1294,116 @@ ngx_http_lua_ffi_set_priv_key(ngx_http_request_t *r,
+
+    if (depth < 0) {
+        sscf = ngx_http_get_module_srv_conf(r, ngx_http_ssl_module);
+        if (sscf != NULL) {
```
we'd better set the `depth` to 1 when `sscf` is `NULL`, and a comment: `same as the default value of ssl_verify_depth`.

ArchangelSDY replied:
sure. updated.

rainingmaster commented:
@ArchangelSDY 
Hello, I see the [default cert](https://github.com/openresty/lua-nginx-module/pull/1666/files#diff-91e07e8def01efe511547241b8afaea8R861) is same with the [cert loaded in ffi api](https://github.com/openresty/lua-nginx-module/pull/1666/files#diff-91e07e8def01efe511547241b8afaea8R842), maybe we should change another cert as new CA cert? I also comment [here](https://github.com/openresty/lua-resty-core/pull/289#issuecomment-616292311)

zhuizhuhaomeng commented on src/ngx_http_lua_ssl_certby.c:
```diff
@@ -1294,4 +1294,119 @@ ngx_http_lua_ffi_set_priv_key(ngx_http_request_t *r,
+
+
+int
+ngx_http_lua_ffi_ssl_verify_client(ngx_http_request_t *r, void *ca_certs,
```
if arg ca_certs is path of the cert, then the code would be must easizer.  just like ngx_ssl_client_certificate
why not use path ?

ArchangelSDY replied:
Since `ssl.set_cert` accepts a pointer rather than a path, I think it should better to keep them consistent. Besides, it provides some flexibility that these certificates do not have to be on disk. For example, in [Kubernetes ingress controller](https://github.com/kubernetes/ingress-nginx) use case, certificates are kept in Nginx memory and are update via a REST API.

ArchangelSDY commented:
@rainingmaster  I see your concern. Then I will use a different cert at server side. Now both certs are self signed so the CA lists should not interfere with each other.

rainingmaster commented:
> @rainingmaster I see your concern. Then I will use a different cert at server side. Now both certs are self signed so the CA lists should not interfere with each other.

@ArchangelSDY Good job, and I think `Test 8` compare to `Test 6` should only difference in *client provides no certificate*, but should provide CA cert [here](https://github.com/openresty/lua-nginx-module/pull/1666/files#diff-91e07e8def01efe511547241b8afaea8R983), like `Test 7` should only different with `Test 6` in *CA certificates*?

Beside, `server_tokens` and `lua_ssl_trusted_certificate` should be useless in new test cases;

rainingmaster commented on src/ngx_http_lua_ssl_certby.c:
```diff
@@ -1294,4 +1294,119 @@ ngx_http_lua_ffi_set_priv_key(ngx_http_request_t *r,
+        *err = "bad request";
+        return NGX_ERROR;
+    }
+
```
As [the document describe](https://github.com/openresty/lua-resty-core/pull/289/files#diff-a69ae86a1de8bc9594d4d4f6dca89709R482), here should check the phase like [ngx.exit](https://github.com/openresty/lua-nginx-module/blob/master/src/ngx_http_lua_control.c#L363). And now the function can be run other phase such as `content_by_lua_block`.

ArchangelSDY replied:
I see other SSL APIs don't have this check either. Is this simply because of some historical reasons?

ArchangelSDY replied:
Check added.

rainingmaster replied:
> I see other SSL APIs don't have this check either. Is this simply because of some historical reasons?

Yes, I agree, but I think we should fix these inconsistencies(such as [set_der_cert](https://github.com/openresty/lua-resty-core/blob/master/lib/ngx/ssl.md#set_der_cert)) with the documentation and add some test cases, right? @doujiang24

ArchangelSDY commented:
> > @rainingmaster I see your concern. Then I will use a different cert at server side. Now both certs are self signed so the CA lists should not interfere with each other.
> 
> @ArchangelSDY Good job, and I think `Test 8` compare to `Test 6` should only difference in _client provides no certificate_, but should provide CA cert [here](https://github.com/openresty/lua-nginx-module/pull/1666/files#diff-91e07e8def01efe511547241b8afaea8R983), like `Test 7` should only different with `Test 6` in _CA certificates_?
> 
> Beside, `server_tokens` and `lua_ssl_trusted_certificate` should be useless in new test cases;

Make sense. Updated `Test 8` and removed those useless parts.

spacewander commented on src/ngx_http_lua_ssl_certby.c:
```diff
@@ -1367,7 +1367,7 @@ ngx_http_lua_ffi_ssl_verify_client(ngx_http_request_t *r, void *ca_certs,
+        ca_store = X509_STORE_new();
         if (ca_store == NULL) {
             *err = "SSL_CTX_get_cert_store() failed";
             return NGX_ERROR;
```
Remember to `goto failed` when failed to create `name_chain`:
https://github.com/openresty/lua-nginx-module/blob/6b58760fe4da41c895ab8dc4ddbd8c5feccc75f2/src/ngx_http_lua_ssl_certby.c#L1378-L1381

ArchangelSDY replied:
Fixed.

rainingmaster commented:
@agentzh @doujiang24 hihi, since 1.17.8 is released, could we merge this PR now?

ArchangelSDY commented on t/140-ssl-c-api.t:
```diff
@@ -812,3 +816,206 @@ lua ssl server name: "test.com"
+            local cert_data = f:read("*all")
+            f:close()
+
+            local cert = ffi.C.ngx_http_lua_ffi_parse_pem_cert(cert_data, #cert_data, errmsg)
```
I see. Just fixed them. For the second leak I think it should be caused by SSL session reuse. After I set `proxy_ssl_session_reuse` to `off`, it looks good now.

doujiang24 commented:
@ArchangelSDY Thanks for your contribution again.
The new OpenResty 1.19.3.1 release is coming, and we are preparing an online open meeting to introduce the new release.
Do you have time to have a talk about this new feature?
My email is doujiang24#gmail.com and my WeChat ID is doujiang24, welcome to contact me if you are interested in it.
